# STA130 HW07 — 14 November, 2024

### Andrew Pols 

___

# 1. Conceptual Definitions

### 1.1 The difference between Simple Linear Regression and Multiple Linear Regression; and the benefit the latter provides over the former:

Multiple Linear Regression (MLR) extends the concept of Simple Linear Regression by incorporating multiple predictor variables to model the outcome variable. The model is represented by an equation similar to the SLR model, but with additional terms for each variable:

$$ \large Y_i = \beta_0 + \beta_1 x_1 + \beta_2 x_2 + \cdots + \beta_i x_i$$

Here, we interpret $\beta_0$ as the intercept of the line modelling the **continuous** outcome variable. (Note that in alignment with the course notes, we will assume the $\epsilon$ term to be present without writing it, for simplicity).

MLR can offer a more realistic model by considering multiple factors that might influence an outcome; it can be unreasonable to assume just ONE predictor influences the outcome, so MLR takes the idea of SLR and extends it to multiple influences, and allows us to compare the strengths of those influences on the outcome.

We will discuss further categorical and continuous variations of the above equation in the following answers.


$$ \large Y_i = \beta_0 + \beta_1 1_\text{[group]}k_i + \beta_2 1_\text{[group]}k_i + \cdots + \beta_i 1_\text{[group]}k_i $$ 

__

### 1.2: the difference between using a continuous variable and an indicator variable in Simple Linear Regression; and these two linear forms:

The continuous form of Simple Linear Regression takes on the equation:

$$ \large Y_i = \beta_0 + \beta_1 x_1$$

The categorical indicator variable form of SLR takes on the equation:

$$ \large Y_i = \beta_0 + \beta_1 \times 1_\text{[group]} k_i $$

In the continuous variable model, the slope coefficient $\beta_1$ is the average change in the outcome variable for a one-unit increase in the **continuous** predictor variable. The $\beta_0$ term is the y-intercept of the line that is modelled; this is simply a vertical translation of the line, or can also be thought of as the value of $Y_i$ when $x_i = 0$.

In the indicator model, we deal with categorical predictor variables instead of continuous ones. Thus, we are no longer dealing in terms of slopes and intercepts, but baselines and contrasts. The baseline of the model is $\beta_0$, whereas the contrast is $\beta_1$. For a categorical variable, $k$, if $k_i$ satisfies the desired group (ie. $k_i \in \text{group}$), then the contrast variable $\beta_1$ is "turned on." So, the average of such variables is $\beta_0 + \beta_1$. Conversely, the average of predictor variables that don't satisfy the group condition have an average of $\beta_0$. So, the difference between these two averages is $\beta_1$; hence, the name "contrast" variable. 

So, the indicator model helps find the difference in categories of a predictor variable.

__

### 1.3: the change that happens in the behavior of the model (i.e., the expected nature of the data it models) when a single indicator variable is introduced alongside a continuous variable to create a Multiple Linear Regression; and these two linear forms (i.e., the Simple Linear Regression versus the Multiple Linear Regression):

When modelling an indicator variable with a continuous variable, we obtain the following equation:

$$ \large Y_i = \beta_0 + \beta_1 \times 1_\text{[group]} k_i + \beta_2 z_i $$

In this model, we have a combined explanation of categorical comparison and continuous effects. That is, we still retain the same properties in which $\beta_0$ is the ***baseline***, but now we have two different types of coefficients. In the aboce equation, we have a categorical variable $k_i$, in which a $k_i$ satisfying a given group is applied a contrast of $\beta_1$. However, we now also have a slope coefficient of $\beta_2$. 

This means that we form two parallel lines with the same slope $\beta_2$. The lines will differ by a vertical translation of $\beta_1$, since this is the contrast between the two groups being studied. We can extend this idea to more variables, as outlined below.

### 1.4: the effect of adding an interaction between a continuous and an indicator variable in Multiple Linear Regression models; and this linear form:

The model in 1.3 accounts for a constant slope between groups, producing parallel lines. But, what if we wanted to see the continuous effect of a group on the outcome? We may add the below interaction:

$$ \large Y_i = \beta_0 + \beta_1 x_i + \beta_2 \times 1_{[\text{group}]}(k_i) + \beta_3 \times x_i \times 1_{[\text{group}]}(k_i)$$

Thus, by adding the indicator "switch" to a multiplication of a slope coefficient, we can adjust each group's slope accordingly. 

For variables not satisfying the group condition (ie. $k_i = 0$), we get $Y_i = \beta_0 + \beta_1 x_i$, 
Otherwise, for variables satisfying the group condition, we get $ Y_i = \beta_0 + \beta_1 x_i + \beta_2 \times 1_{[\text{group}]}(k_i) + \beta_3 \times x_i \times 1_{[\text{group}]}(k_i)$.

The first line has intercept $\beta_0$ and slope $\beta_1$. The second line has intercept $\beta_0 + \beta_2$ and slope $\beta_1 + \beta_3$.

Thus, this interaction allows us to have differnet slopes depending on the group term, providing a dynamic modelling of predictors to outcomes.

### 1.5: the behavior of a Multiple Linear Regression model (i.e., the expected nature of the data it models) based only on indicator variables derived from a non-binary categorical variable; this linear form; and the necessarily resulting binary variable encodings it utilizes:

The previous answers addressed binary categorical variables, in which we either satisfy a condition, or we don't. For categories with multiple responses, we can actually employ the same methodology. So, for one "group" of categories, we have one choice of the group as the baseline, with the rest of the group as the references to that baseline (meaning we have to add indicator variables to make this reference). 

The following is the linear form of the model:

$$Y_i = \beta_0 + \sum_{j=1}^{K-1} \beta_j \times 1_{[\text{Category j}]}(k_i)$$

As we see below, we have K-1 indicator variables, where K is the number of categories in a "group" of predictor variables. The "unaccounted" variable is represented by $\beta_0$. Thus, every other category is a reference to this baseline. This is done by this "binary encoding" that flips an associated $k_i$ **on or off** (binary) depending on the satisfaction of a condition.


## NotebookLM Session Summary




The conversation focused on understanding simple and multiple linear regression and the use of continuous and indicator variables in both.  It began by comparing SLR, which models the relationship between one predictor and an outcome, to MLR, which allows for multiple predictors. The discussion then explored the use of continuous variables as predictors in both SLR and MLR. The conversation then turned to non-binary categorical variables, and how they can be represented in MLR using indicator variables, each representing a specific category except for the baseline category. This representation helps to understand the differences in the outcome variable across various categories relative to the chosen baseline. The coefficients associated with these indicator variables then quantify the average difference in the outcome variable between the represented category and the baseline. To provide a more complete mathematical understanding, the conversation also touched upon the generalized form of MLR, both with sigma notation, which provides a precise mathematical representation, and without sigma notation for a more intuitive understanding. The conversation concluded by using a concrete example of house prices and regions to illustrate these concepts.
 


___

# 2. Multiple Linear Regression Example Discussion

### 2.1 

Let Sales be the outcome variable, predicted by continuous variables TV and Online, which we will call $x$ and $z$ respectively.

Without any interaction between variables, we obtain the following model from the scenario:

$$ \text{Sales} = \beta_0 + \beta_1 x_i + \beta_2 z_i $$

Thus, we observe that sales are predicted independently and ***continuously*** by each predictor variable. That is, the coefficients of each predictor is unaffected by the other. 

However, if we wanted an interaction between the two, we could have a "weighted" affect by adding an additional term that quantifies the combined effect of both predictor variables at the same time:

$$ \text{Sales} = \beta_0 + \beta_1 x_i + \beta_2 z_i + \beta_3 x_i z_i$$

Intuitively, we can characterize this 3rd term as being a syngergistic effect between the two variables; if $x_i$ is large, but $z_i$ is small, then the effect of $x_i$ is diminished (and vice versa). But if both variables are large, their combined effect will be even greater.

### 2.2

In the case of categorical variables, we may categorize the predictor variables as being "High" or "Not High" (ie. low) for the purpose of creating binary varibales. The linear form of this model becomes:

$$ \large \text{Sales} = \beta_0 + \beta_1 \times 1_\text{[High]} x_i + \beta_2 \times 1_\text{[High]}z_i$$

In this case, $\beta_0$ is the baseline, which occurs when TV and Online budgets are BOTH low. 

When TV budget is high, but Online budgets are low, we add the contrast $\beta_1$, for an average of $\beta_0 + \beta_1$. 

When Online Budget is high, but TV budget is low, we add the contrast $\beta_2$, for an average of $\beta_0 + \beta_2$. 


However, this isnt particularly nuanced, as each variable is independent and does not affect the other. We can add a synergestic interaction between them by the following equation:

$$ \large \text{Sales} = \beta_0 + \beta_1 \times 1_\text{[High]} x_i + \beta_2 \times 1_\text{[High]}z_i + \beta_3 \times 1_\text{[High]} x_i z_i$$.

Similarly to the explanation in 2.1, the effect here is synergestic in the last term. This means that the effects of the variables are combined. Since we're dealing with categorical variables, for the last term to be activated BOTH TV and Online budgets must be high, but this model allows us to see WHAT effect this has in this case (which is intuitively more telling...).


__

## NotebookLM Session Summary
This conversation began by analyzing a hypothetical scenario of a company selling sports equipment and examining the relationship between sales and advertising spending on TV and online platforms.  The initial focus was on identifying the outcome variable (sales) and the predictor variables (TV and online advertising spending), and recognizing the potential for an interaction effect between these predictors. The discussion then shifted to formulating linear models, both with and without interaction terms, to represent the relationship between the variables. These linear models were expressed using mathematical formulas, considering both continuous and categorical representations of the predictor variables. For example, with continuous predictors, *Sales* = β<sub>0</sub> + β<sub>1</sub>&#x20; *TV* + β<sub>2</sub>&#x20; *Online* represents a model without interaction, while *Sales* = β<sub>0</sub> + β<sub>1</sub>&#x20; *TV* + β<sub>2</sub>&#x20; *Online* + β<sub>3</sub>&#x20; *TV* \* *Online* incorporates an interaction term. A key point of clarification involved the distinction between multiple predictor variables and indicator variables, emphasizing that indicator variables are specifically used to represent categorical variables in regression models. The conversation concluded by reiterating the importance of selecting the appropriate type and number of variables (continuous or indicator) based on the nature of the data and the relationships being modeled.  

___

# 3. CSCH Logistic Regression

### 3.1.

Firstly, we will create an additive linear (logistic) regression based on the following variables.

**Outcome:** WELLNESS_self_rated_mental_health_binary (a binary variable which we encode to either be "good" or "bad").

**Predictors**: DEMO_age (continuous age variable), CONNECTION_community_barriers_introvert_binary (binary indicator variable of self-reported introvert vs. extrovert), CONNECTION_community_involvement_sports_p3m_binary (binary indicator variable reporting a person's sports activity level).

If this were linear regression, it would look like this:

$$ \large Y_i = \beta_0 + \beta_1 z_\text{age} + \beta_2 \times 1_\text{["introvert"]}k_i + \beta_3 \times 1_\text{["sporty"]}k_i$$

The below code sets up and creates our model.

In [63]:
import pandas as pd
cols = pd.read_csv("../course_project/data/var_names.csv")

data = pd.read_csv("../course_project/data/CSCS_data_anon.csv", 
                   na_values=["9999", "", " ", "Presented but no response", "NA"],
                   low_memory=False)  # Specify the desired type


empty = (data.isna().sum()==data.shape[0])
data = data[empty.index[~empty]] # keep non empty columns only

data = data[data.REMOVE_case=='No']

data = data[data.SURVEY_collection_year==2022]

data.shape

(3896, 1779)

In [71]:
import pandas as pd
import statsmodels.formula.api as smf
import numpy as np

# Binary mapping for WELLNESS_self_rated_mental_health
data.loc[:, 'WELLNESS_self_rated_mental_health_binary'] = data['WELLNESS_self_rated_mental_health'].map({
    'Poor': 0,
    'Fair': 0,
    'Good': 1,
    'Very good': 1,
    'Excellent': 1
})

# Define binary mapping for the introvert indicator
introvert_mapping = {
    'I am an introvert': 1,
    'Not Selected': 0,
    np.nan: 0  # Convert NaN values to 0
}

# Apply mapping to create binary column
data.loc[:, 'CONNECTION_community_barriers_introvert_binary'] = \
    data['CONNECTION_community_barriers_introvert'].map(introvert_mapping)

# Define the mapping for sports involvement
sports_mapping = {
    "Yes": 1,
    "No, but I would like to": 0,
    "No, and I'm not interested in doing so": 0,
    np.nan: 0  # Treat NaNs as "not interested"
}

# Apply the mapping to create the binary column
data.loc[:, 'CONNECTION_community_involvement_sports_p3m_binary'] = data['CONNECTION_community_involvement_sports_p3m'].map(sports_mapping)

# Ensure there are no NaNs in the outcome variable and predictors
data = data.dropna(subset=['WELLNESS_self_rated_mental_health_binary', 'CONNECTION_community_involvement_sports_p3m_binary', 'CONNECTION_community_barriers_introvert_binary', 'DEMO_age'])

In [70]:

# Define the logistic regression formula
formula_additive = 'WELLNESS_self_rated_mental_health_binary ~ DEMO_age + CONNECTION_community_barriers_introvert_binary + CONNECTION_community_involvement_sports_p3m_binary'

# Fit the logistic regression model
model_additive = smf.logit(formula=formula_additive, data=data).fit()

# Print the model summary
print(model_additive.summary())


Optimization terminated successfully.
         Current function value: 0.626938
         Iterations 6
                                      Logit Regression Results                                      
Dep. Variable:     WELLNESS_self_rated_mental_health_binary   No. Observations:                 3148
Model:                                                Logit   Df Residuals:                     3144
Method:                                                 MLE   Df Model:                            3
Date:                                      Fri, 15 Nov 2024   Pseudo R-squ.:                 0.01182
Time:                                              02:44:50   Log-Likelihood:                -1973.6
converged:                                             True   LL-Null:                       -1997.2
Covariance Type:                                  nonrobust   LLR p-value:                 3.107e-10
                                                         coef    std err          z      P

### 3.2.
Leaving the above for later interpretation, we move on to a synergistic form of regression.

The form of this model will have an interaction between (1) age and reported introversion, as well as (2) age and reported sport activity.

The below equation models this:

$$ \large Y_i = \beta_0 + \beta_1 z_{\text{age}} + \beta_2 \times 1_{\text{["introvert"]}}k_i + \beta_3 \times 1_{\text{["sporty"]}}k_i + \beta_4 \left( z_{\text{age}} \times 1_{\text{["introvert"]}}k_i \right) + \beta_5 \left( z_{\text{age}} \times 1_{\text{["sporty"]}}k_i \right)
$$

In [72]:
# Synergistic model formula with main effects and interaction terms
formula_synergistic = (
    'WELLNESS_self_rated_mental_health_binary ~ '
    'DEMO_age + CONNECTION_community_barriers_introvert_binary + '
    'CONNECTION_community_involvement_sports_p3m_binary + '
    'DEMO_age:CONNECTION_community_barriers_introvert_binary + '
    'DEMO_age:CONNECTION_community_involvement_sports_p3m_binary'
)

# Fit the model
model_synergistic = smf.logit(formula=formula_synergistic, data=data).fit()

# Display the model summary
print(model_synergistic.summary())


Optimization terminated successfully.
         Current function value: 0.626839
         Iterations 6
                                      Logit Regression Results                                      
Dep. Variable:     WELLNESS_self_rated_mental_health_binary   No. Observations:                 3148
Model:                                                Logit   Df Residuals:                     3142
Method:                                                 MLE   Df Model:                            5
Date:                                      Fri, 15 Nov 2024   Pseudo R-squ.:                 0.01198
Time:                                              02:45:12   Log-Likelihood:                -1973.3
converged:                                             True   LL-Null:                       -1997.2
Covariance Type:                                  nonrobust   LLR p-value:                 3.807e-09
                                                                  coef    std err         

### 3.3.

As noted above, the additive model is as follows:
$ \;\; Y_i = \beta_0 + \beta_1 z_\text{age} + \beta_2 \times 1_\text{["introvert"]}k_i + \beta_3 \times 1_\text{["sporty"]}k_i$

**Intuitive Explanation:**
- $\beta_0$: The intercept, or baseline, outcome when age = 0, and a person is not introverted or sporty.
- $\beta_1$: The slope coefficient of age; measures the rate of change of the outcome for each 1-year increase in age.
-$\beta_2$: The contrast term between introverts and extroverts, levaraging binary encoding (purely additive; has no effect on age/sportiness).
- $\beta_3$: The contrast term between sporty vs non-sporty people.

__

As noted above, the synergistic model is as follows:
$ Y_i = \beta_0 + \beta_1 z_{\text{age}} + \beta_2 \times 1_{\text{["introvert"]}}k_i + \beta_3 \times 1_{\text{["sporty"]}}k_i + \beta_4 \left( z_{\text{age}} \times 1_{\text{["introvert"]}}k_i \right) + \beta_5 \left( z_{\text{age}} \times 1_{\text{["sporty"]}}k_i \right)
$

**Intuition:**
- This has the same interpretation as the additive model, but also assumes interaction between predictors. 
- As explained in Q2, these interactions result in one predictor having a dependent impact on the effect of another. 
- In the synergistic model, being both "sporty" and "introverted" might have a stronger or weaker combined effect, not just the sum of their individual effects.

### 3.4.
In the additive model, we obtain the following results:

- DEMO_age seems to have a slightly negative slope-coefficient of -0.0040, but with a p-value of 0.069, there is no evidence to suggest this is not just random chance (ie. can't reject $H_0$).


- CONNECTION_community_barriers_introvert_binary seems to have a negative contrast term of -0.8260. This means that, on average, introverts tend to be slightly less content in terms of their mental health compared to the extrovert baseline. With a p-value of 0.00, we have extremely strong evidence against $H_0$. 


- CONNECTION_community_involvement_sports_p3m_binary seems to have a positive contrast of 1.2777. This means that, on average, sporty people tend to be slightly more content in their mental health compared to the non-sporty baseline. With a p-value of 0.00, we have extremely strong evidence against $H_0$. 


In the synergistic model, we obtain the following results:
- DEMO_age, CONNECTION_community_barriers_introvert_binary, and CONNECTION_community_involvement_sports_p3m_binary on their own have about the same coefficient values.
    - However, since we've added the effect of more interactions, the statistical evidence of the latter two predictors has decreased, with o-values of 0.032 and 0.037 respectively (although this is still enough evidence to reject $H_0$).

<br>

- For the interaction between DEMO_age:CONNECTION_community_barriers_introvert_binary, we observe a slightly positive coefficient of 0.0070. That is, on average, as age increases by one year, the log odds of the outcome occurring increase by 0.0070 more than they would for non-introverts.
    - However, with a p-value of 0.506, we cannot reject $H_0$ — not statistically significant.

<br>


- For the interaction between DEMO_age:DEMO_age:CONNECTION_community_involvement_sports_p3m_binary, we observe a slightly negative coefficient of -0.0081. That is, on average, as age increases by one year, the log odds of the outcome occurring decrease by -0.0080 more than they would for non-sporty people.
    - However, with a p-value of 0.625, we cannot reject $H_0$ — not statistically significant.


### 3.5.

In [77]:
import plotly.graph_objects as go

# Additive model plot
fig_additive = go.Figure()

# Plot the data points for the outcome (1s and 0s)
fig_additive.add_trace(go.Scatter(x=data['DEMO_age'], y=data['WELLNESS_self_rated_mental_health_binary'], 
                                 mode='markers', name='Actual Data (Additive)',
                                 marker=dict(color='blue', size=8)))

# Add predicted probabilities for the additive model
fig_additive.add_trace(go.Scatter(x=data['DEMO_age'], y=model_additive.fittedvalues, 
                                 mode='markers', name='Predicted Probabilities (Additive)',
                                 marker=dict(color='green', size=8)))

# Add titles and axis labels
fig_additive.update_layout(
    title="Additive Model - Age vs. Mental Health",
    xaxis_title="Age",
    yaxis_title="Predicted Probability of Higher Mental Health",
    showlegend=True
)

fig_additive.show()


In [78]:
# Synergistic model plot
fig_synergistic = go.Figure()

# Plot the data points for the outcome (1s and 0s)
fig_synergistic.add_trace(go.Scatter(x=data['DEMO_age'], y=data['WELLNESS_self_rated_mental_health_binary'], 
                                    mode='markers', name='Actual Data (Synergistic)',
                                    marker=dict(color='blue', size=8)))

# Add predicted probabilities for the synergistic model
fig_synergistic.add_trace(go.Scatter(x=data['DEMO_age'], y=model_synergistic.fittedvalues, 
                                    mode='markers', name='Predicted Probabilities (Synergistic)',
                                    marker=dict(color='red', size=8)))

# Add titles and axis labels
fig_synergistic.update_layout(
    title="Synergistic Model - Age vs. Mental Health",
    xaxis_title="Age",
    yaxis_title="Predicted Probability of Higher Mental Health",
    showlegend=True
)

fig_synergistic.show()


In [79]:
# Overlay plot with both additive and synergistic models
fig_overlay = go.Figure()

# Plot the data points for the outcome (1s and 0s)
fig_overlay.add_trace(go.Scatter(x=data['DEMO_age'], y=data['WELLNESS_self_rated_mental_health_binary'], 
                                mode='markers', name='Actual Data', marker=dict(color='blue', size=8)))

# Add predicted probabilities for the additive model
fig_overlay.add_trace(go.Scatter(x=data['DEMO_age'], y=model_additive.fittedvalues, 
                                mode='markers', name='Predicted Probabilities (Additive)', 
                                marker=dict(color='green', size=8)))

# Add predicted probabilities for the synergistic model
fig_overlay.add_trace(go.Scatter(x=data['DEMO_age'], y=model_synergistic.fittedvalues, 
                                mode='markers', name='Predicted Probabilities (Synergistic)', 
                                marker=dict(color='red', size=8)))

# Add titles and axis labels
fig_overlay.update_layout(
    title="Overlay of Additive and Synergistic Models - Age vs. Mental Health",
    xaxis_title="Age",
    yaxis_title="Predicted Probability of Higher Mental Health",
    showlegend=True
)

fig_overlay.show()


### Interpretation

The plot shows the relationship between age and the probability of having a higher self-rated mental health score (binary outcome: 1 for higher mental health, 0 for lower). 

- **Actual Data**: The blue markers represent the observed outcomes (1s and 0s) for each individual's age.
- **Additive Model**: The green markers indicate the predicted probabilities based on the additive logistic regression model, which considers age and introversion status separately.
- **Synergistic Model**: The red markers represent the predicted probabilities from the synergistic model, which includes an interaction term between age and introversion status.

The predicted probabilities are continuous values between 0 and 1, reflecting the likelihood of a higher self-rated mental health score as age increases, under both models. The synergy between age and introversion status is captured more explicitly in the red markers.

__
## ChatGPT Session Summary
https://chatgpt.com/share/6736b7d7-5668-800e-a35a-afe6ce05b972

Over the course of our interactions, we discussed the process of fitting logistic regression models with binary predictors and visualizing the results. The focus was on interpreting additive and synergistic models, where the outcome was a binary mental health rating, influenced by age and introversion status. After addressing various issues with model fitting and plot formatting, I provided a step-by-step approach to visualize the data, first through individual plots for each model, followed by an overlay of both models for comparison. The discussion involved ensuring the plots used points instead of lines and involved making appropriate adjustments to clarify the results, focusing on simplicity and clarity for interpretation.

___

# 4. Statistical Evidence: R-Squared vs. P-Values

Consider the following code.

In [83]:
import pandas as pd

url = "https://raw.githubusercontent.com/KeithGalli/pandas/master/pokemon_data.csv"
# fail https://github.com/KeithGalli/pandas/blob/master/pokemon_data.csv
pokeaman = pd.read_csv(url) 
pokeaman


#                   Name   Type 1  Type 2  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison  45      49       49   
1      2                Ivysaur    Grass  Poison  60      62       63   
2      3               Venusaur    Grass  Poison  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison  80     100      123   
4      4             Charmander     Fire     NaN  39      52       43   
..   ...                    ...      ...     ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark  80     160       60   
799  721              Volcanion     Fire   Water  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0         65       65     45           1      False  
1         80       80     60           1      False  
2        100      100     80           1      False  
3        122      120     80           1      False  
4         60       50     65           1      False  
..       ...      ...    ...         ...        ...  
795      100      150     50           6       True  
796      160      110    110           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True  

[800 rows x 12 columns]

In [84]:
import statsmodels.formula.api as smf

model1_spec = smf.ols(formula='HP ~ Q("Sp. Def") + C(Generation)', data=pokeaman)
model2_spec = smf.ols(formula='HP ~ Q("Sp. Def") + C(Generation) + Q("Sp. Def"):C(Generation)', data=pokeaman)
model2_spec = smf.ols(formula='HP ~ Q("Sp. Def") * C(Generation)', data=pokeaman)

model2_fit = model2_spec.fit()
model2_fit.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     HP   R-squared:                       0.176
Model:                            OLS   Adj. R-squared:                  0.164
Method:                 Least Squares   F-statistic:                     15.27
Date:                Fri, 15 Nov 2024   Prob (F-statistic):           3.50e-27
Time:                        02:55:19   Log-Likelihood:                -3649.4
No. Observations:                 800   AIC:                             7323.
Df Residuals:                     788   BIC:                             7379.
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
===================================================================================================
                                      coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------------------
Intercept                          26.8971      5.246      5.127      0.000      16.599      37.195
C(Generation)[T.2]                 20.0449      7.821      2.563      0.011       4.692      35.398
C(Generation)[T.3]                 21.3662      6.998      3.053      0.002       7.629      35.103
C(Generation)[T.4]                 31.9575      8.235      3.881      0.000      15.793      48.122
C(Generation)[T.5]                  9.4926      7.883      1.204      0.229      -5.982      24.968
C(Generation)[T.6]                 22.2693      8.709      2.557      0.011       5.173      39.366
Q("Sp. Def")                        0.5634      0.071      7.906      0.000       0.423       0.703
Q("Sp. Def"):C(Generation)[T.2]    -0.2350      0.101     -2.316      0.021      -0.434      -0.036
Q("Sp. Def"):C(Generation)[T.3]    -0.3067      0.093     -3.300      0.001      -0.489      -0.124
Q("Sp. Def"):C(Generation)[T.4]    -0.3790      0.105     -3.600      0.000      -0.586      -0.172
Q("Sp. Def"):C(Generation)[T.5]    -0.0484      0.108     -0.447      0.655      -0.261       0.164
Q("Sp. Def"):C(Generation)[T.6]    -0.3083      0.112     -2.756      0.006      -0.528      -0.089
==============================================================================
Omnibus:                      337.229   Durbin-Watson:                   1.505
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             2871.522
Skew:                           1.684   Prob(JB):                         0.00
Kurtosis:                      11.649   Cond. No.                     1.40e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.4e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

## Conceptual Explanations

We often treat p-values and $R^2$ as similar evidences of a combined purpose in the context of regression analysis. However, it is not true that both of them tell the same story, or that they both must have a "statistically significant" result to "confirm" each other. 


The apparent contradiction between the two statements comes from the different things that **R-squared** and **p-values** represent in regression analysis.

**R-squared** shows how much of the variability in the outcome variable is explained by the model. In this case, a value of **17.6%** means that the model explains only a small portion of the total variability in the data. While the model does capture some key relationships, it suggests that other factors not included in the model may also play a significant role in affecting the outcome.

**P-values**, on the other hand, test whether the effect of each predictor variable is statistically significant. A **low p-value** indicates that there is strong evidence against the null hypothesis (which suggests no effect), meaning the predictor has a **meaningful relationship** with the outcome, controlling for other variables in the model.

However, ***these two concepts ARE NOT MUTUALLY EXCLUSIVE***.

These concepts don't conflict; instead, they focus on different aspects of the model:

- **R-squared** tells us about the overall goodness of fit—**how well the model explains the variation in the outcome**. 

- **P-values**, however, show us the strength of evidence for the relationship between each predictor and the outcome, **independent of the model’s overall explanatory power**.

It’s completely possible for a model to have a low **R-squared** (indicating limited explanatory power) and still have **statistically significant coefficients** with **large effect sizes**. This means that while the predictors in the model have **strong individual effects** on the outcome, there could be other factors outside of the model that account for a significant portion of the outcome's variability. So, **both measures can be interpreted meaningfully together** because they focus on different aspects of the model's performance and the predictors' relationships with the outcome.

## Example Explanation

- **R-squared = 17.6%** means the model explains only a small portion (17.6%) of the variability in the outcome, **suggesting there are other unmeasured factors** contributing to the remaining 82.4% of the variation.

- **Large coefficients with low p-values** indicate that the predictors included in the model have a **significant relationship** with the outcome and **contribute meaningfully** to predicting the outcome variable.

In summary, the model captures some of the variability in the outcome, but **not enough to fully explain it**. Despite the low **R-squared**, the predictors are still **important** and have a **strong effect** on the outcome, as shown by their **large coefficients** and **low p-values**.

This isn't contradictory because **R-squared** and **p-values** measure different aspects of the model: **R-squared** shows how much of the outcome's variability is explained by the model, while **p-values** indicate the significance of each predictor's relationship with the outcome, regardless of the model's overall explanatory power.

__

## NotebookLM Session Summary
The discussions focused on clarifying the relationship between p-values and R-squared in statistical models. While not contradictory, these metrics address different aspects of model evaluation. P-values assess the statistical significance of individual predictor variables, indicating the probability of observing the estimated coefficient if there were no real association. R-squared, on the other hand, measures the model's overall goodness of fit, representing the proportion of variation in the outcome variable explained by the model. It is possible to have statistically significant predictors (low p-values) even with a low R-squared, suggesting that while individual predictors may have a real effect, the model as a whole does not account for much of the outcome variability. This can occur due to omitted variables, weak relationships, or a small sample size. The relative importance of p-values and R-squared depends on the analysis goals, with prediction tasks often prioritizing R-squared and causal inference focusing on p-values and coefficient interpretation. Ultimately, both metrics are important for a balanced evaluation of a model's strengths and limitations. 

___

# 5. Regression and $R^2$ Revisited

Consider the five cells of code below, for which I will breifly describe what each of them is doing, and then provide a final overall explanation at the end.

## Cell 1
In this step, the dataset (`pokeaman`) is divided into two parts: a training set and a testing set, with each receiving 50% of the data. To ensure smooth analysis, missing values in the "Type 2" column are replaced with the string `'None'`, avoiding errors due to NaN values. The `train_test_split` function is used to separate the data, which allows the model to be trained on one part of the data and tested on another. This split is crucial for evaluating how well the model can generalize to new, unseen data, giving a better understanding of its overall performance.

In [86]:
import numpy as np
from sklearn.model_selection import train_test_split

fifty_fifty_split_size = int(pokeaman.shape[0]*0.5)

# Replace "NaN" (in the "Type 2" column with "None")
pokeaman.fillna('None', inplace=True)

np.random.seed(130)
pokeaman_train,pokeaman_test = \
  train_test_split(pokeaman, train_size=fifty_fifty_split_size)
pokeaman_train

#                     Name   Type 1   Type 2  HP  Attack  Defense  \
370  338                  Solrock     Rock  Psychic  70      95       85   
6      6                Charizard     Fire   Flying  78      84       78   
242  224                Octillery    Water     None  75     105       75   
661  600                    Klang    Steel     None  60      80       95   
288  265                  Wurmple      Bug     None  45      45       35   
..   ...                      ...      ...      ...  ..     ...      ...   
522  471                  Glaceon      Ice     None  65      60      110   
243  225                 Delibird      Ice   Flying  45      55       45   
797  720      HoopaHoopa Confined  Psychic    Ghost  80     110       60   
117  109                  Koffing   Poison     None  40      65       95   
409  373  SalamenceMega Salamence   Dragon   Flying  95     145      130   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
370       55       65     70           3      False  
6        109       85    100           1      False  
242      105       75     45           2      False  
661       70       85     50           5      False  
288       20       30     20           3      False  
..       ...      ...    ...         ...        ...  
522      130       95     65           4      False  
243       65       45     75           2      False  
797      150      130     70           6       True  
117       60       45     35           1      False  
409      120       90    120           3      False  

[400 rows x 12 columns]

## Cell 2

This step defines a linear regression model using the formula `'HP ~ Attack + Defense'`, where `HP` is predicted based on `Attack` and `Defense` from the training data (`pokeaman_train`). The `fit()` function is used to estimate the coefficients of the predictors. After fitting the model, a summary is generated, which provides the coefficients for `Attack` and `Defense`, R-squared (indicating how well the model fits the data), and p-values (to test the statistical significance of the predictors).


In [87]:
model_spec3 = smf.ols(formula='HP ~ Attack + Defense', 
                      data=pokeaman_train)
model3_fit = model_spec3.fit()
model3_fit.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     HP   R-squared:                       0.148
Model:                            OLS   Adj. R-squared:                  0.143
Method:                 Least Squares   F-statistic:                     34.40
Date:                Fri, 15 Nov 2024   Prob (F-statistic):           1.66e-14
Time:                        03:41:09   Log-Likelihood:                -1832.6
No. Observations:                 400   AIC:                             3671.
Df Residuals:                     397   BIC:                             3683.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     42.5882      3.580     11.897      0.000      35.551      49.626
Attack         0.2472      0.041      6.051      0.000       0.167       0.327
Defense        0.1001      0.045      2.201      0.028       0.011       0.190
==============================================================================
Omnibus:                      284.299   Durbin-Watson:                   2.006
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             5870.841
Skew:                           2.720   Prob(JB):                         0.00
Kurtosis:                      20.963   Cond. No.                         343.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

The model's **R-squared** value is 0.148, meaning it explains only 14.8% of the variation in `HP`, which is quite low. This suggests other factors are affecting `HP` that aren’t captured by the model. The **Adjusted R-squared** is slightly lower at 0.143, indicating that adding more predictors doesn’t significantly improve the model's explanatory power.

The **coefficients** show that:
- The **Intercept** is 42.5882, meaning the baseline `HP` when both `Attack` and `Defense` are zero is 42.5882.
- For **Attack**, each additional unit increases `HP` by 0.2472. This predictor is statistically significant with a very low p-value (0.000).
- For **Defense**, each additional unit increases `HP` by 0.1001, and it’s also statistically significant with a p-value of 0.028.

Both **Attack** and **Defense** are statistically significant predictors of `HP`, but the low R-squared value shows the model doesn’t explain a lot of the variation in `HP`, indicating there might be other important factors missing from the model.


## Cell 3

This step calculates both the **"in sample"** and **"out of sample"** R-squared values to compare the model’s performance.

1. **Prediction on Test Data**: The model generates predictions for the test data (`pokeaman_test`), representing how well it predicts unseen `HP` values.

2. **"In sample" R-squared**: This value is calculated from the training data (`pokeaman_train`) and measures how well the model fits the training data. It was previously found to be 0.148.

3. **"Out of sample" R-squared**: This is calculated by comparing the predicted values to the actual `HP` values in the test data. It shows how well the model generalizes to new data.

4. **Print Statements**: The first prints the "in sample" R-squared, and the second prints the "out of sample" R-squared.

### Purpose:
This step helps assess the model’s performance by comparing its fit to the training data ("in sample") versus how well it generalizes to new data ("out of sample"). A large difference between the two could indicate overfitting.

In [88]:
yhat_model3 = model3_fit.predict(pokeaman_test)
y = pokeaman_test.HP
print("'In sample' R-squared:    ", model3_fit.rsquared)
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model3)[0,1]**2)

'In sample' R-squared:     0.14771558304519894
'Out of sample' R-squared: 0.21208501873920738



1. **In sample R-squared**: 0.1477 — The model explains about 14.77% of the variation in `HP` for the training data, which reflects a relatively poor fit.

2. **Out of sample R-squared**: 0.2121 — The model explains about 21.21% of the variation in `HP` for the test data, which is better than its fit to the training data.

### Takeaway:
- The **out of sample** R-squared is higher than the **in sample** R-squared, which is unusual. Typically, we'd expect worse performance on unseen data if the model is overfitting. However, in this case, it suggests the model might generalize better than it fits the training data, potentially due to issues with the representativeness of the training set or other influencing factors.

## Cell 4

This cell sets up and fits a **multiple linear regression model** that includes **interaction terms** between predictors:

1. **Formula**:
   - The model is specified with several predictors (`Attack`, `Defense`, `Speed`, etc.) and their **interaction effects** (e.g., how `Attack` and `Defense` jointly affect `HP`). 
   - The `Q()` function ensures that variables with special characters (like `Sp. Def` and `Sp. Atk`) are properly interpreted.
   - The commented-out part would add complex interactions for categorical variables, but it's avoided due to the large number of combinations that could make the model too complex.

2. **Model Specification**:
   - The model is created using `ols()` from **statsmodels**, which fits an ordinary least squares regression.

3. **Model Fitting**:
   - `model4_spec.fit()` fits the model to the training data.

4. **Model Summary**:
   - `model4_fit.summary()` provides detailed output, including coefficients and diagnostic statistics.

### Purpose:
- This model aims to capture more complex relationships by including both main effects and interactions between predictors, offering a more flexible fit for `HP`.


In [89]:

model4_linear_form = 'HP ~ Attack * Defense * Speed * Legendary'
model4_linear_form += ' * Q("Sp. Def") * Q("Sp. Atk")'
# DO NOT try adding '* C(Generation) * C(Q("Type 1")) * C(Q("Type 2"))'
# That's 6*18*19 = 6*18*19 possible interaction combinations...
# ...a huge number that will blow up your computer

model4_spec = smf.ols(formula=model4_linear_form, data=pokeaman_train)
model4_fit = model4_spec.fit()
model4_fit.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     HP   R-squared:                       0.467
Model:                            OLS   Adj. R-squared:                  0.369
Method:                 Least Squares   F-statistic:                     4.764
Date:                Fri, 15 Nov 2024   Prob (F-statistic):           4.23e-21
Time:                        03:41:20   Log-Likelihood:                -1738.6
No. Observations:                 400   AIC:                             3603.
Df Residuals:                     337   BIC:                             3855.
Df Model:                          62                                         
Covariance Type:            nonrobust                                         
====================================================================================================================================
                                                                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------------------------------------------
Intercept                                                          521.5715    130.273      4.004      0.000     265.322     777.821
Legendary[T.True]                                                   -6.1179      2.846     -2.150      0.032     -11.716      -0.520
Attack                                                              -8.1938      2.329     -3.518      0.000     -12.775      -3.612
Attack:Legendary[T.True]                                         -1224.9610    545.105     -2.247      0.025   -2297.199    -152.723
Defense                                                             -6.1989      2.174     -2.851      0.005     -10.475      -1.923
Defense:Legendary[T.True]                                         -102.4030     96.565     -1.060      0.290    -292.350      87.544
Attack:Defense                                                       0.0985      0.033      2.982      0.003       0.034       0.164
Attack:Defense:Legendary[T.True]                                    14.6361      6.267      2.336      0.020       2.310      26.963
Speed                                                               -7.2261      2.178     -3.318      0.001     -11.511      -2.942
Speed:Legendary[T.True]                                            704.8798    337.855      2.086      0.038      40.309    1369.450
Attack:Speed                                                         0.1264      0.038      3.351      0.001       0.052       0.201
Attack:Speed:Legendary[T.True]                                       5.8648      2.692      2.179      0.030       0.570      11.160
Defense:Speed                                                        0.1026      0.039      2.634      0.009       0.026       0.179
Defense:Speed:Legendary[T.True]                                     -6.9266      3.465     -1.999      0.046     -13.742      -0.111
Attack:Defense:Speed                                                -0.0016      0.001     -2.837      0.005      -0.003      -0.001
Attack:Defense:Speed:Legendary[T.True]                              -0.0743      0.030     -2.477      0.014      -0.133      -0.015
Q("Sp. Def")                                                        -5.3982      1.938     -2.785      0.006      -9.211      -1.586
Legendary[T.True]:Q("Sp. Def")                                    -282.2496    126.835     -2.225      0.027    -531.738     -32.761
Attack:Q("Sp. Def")                                                  0.1094      0.034      3.233      0.001       0.043       0.176
Attack:Legendary[T.True]:Q("Sp. Def")                               12.6503      5.851      2.162      0.031       1.141      24.160
Defense:Q("Sp. Def")    

- **Overfitting Risk**: The model is likely overfitting the training data due to its complexity. It includes a large number of predictors (62 coefficients) and interaction terms that may capture noise rather than true relationships. A significant increase in R² could mean the model is capturing more nuances in the training data, but it could also mean it's overfitting to noise or specific patterns that won’t generalize well to new data.
  
- **Key Indicators**:
  - The model shows a large gap between **R-squared** (0.467) and **Adjusted R-squared** (0.369), suggesting the added complexity isn't improving the model's generalizability.
  - Many predictors, including interaction terms, are statistically significant, but these relationships may not be meaningful beyond the training data.
  - High **AIC** and **BIC** values (3603 and 3855) also signal that the model might be too complex.

- **Possible Solutions**:
  - **Cross-validation** can help assess the model's performance on unseen data.
  - Reducing the number of predictors, particularly interactions, could simplify the model and improve its ability to generalize.

## Cell 5

### In-sample vs Out-of-sample R-squared:
- **In-sample R²** tells you how well the model fits the training data.
- **Out-of-sample R²** tells you how well the model generalizes to unseen data (test set).

### Key Insights:
- **High In-sample R² vs Low Out-of-sample R²** suggests **overfitting**, where the model fits the training data well but struggles with new data.
- **Similar High R² values** in both in-sample and out-of-sample suggest the model generalizes well without overfitting.

If there's a large difference between in-sample and out-of-sample R², the model might be overfitting and could benefit from simplification or regularization.

In [91]:
yhat_model4 = model4_fit.predict(pokeaman_test)
y = pokeaman_test.HP
print("'In sample' R-squared:    ", model4_fit.rsquared)
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model4)[0,1]**2)

'In sample' R-squared:     0.46709442115833855
'Out of sample' R-squared: 0.002485342598992873


The stark contrast between the **in-sample** (0.4671) and **out-of-sample** (0.0025) R-squared values indicates a clear **overfitting** issue:

### Key Points:
- **In-sample R-squared (0.4671)**: The model explains about **47%** of the variance in the training data, which is a moderate fit.
- **Out-of-sample R-squared (0.0025)**: The model performs poorly on the test set, explaining almost no variance. This suggests that it has **failed to generalize** to new data.

### Conclusion:
- The **large discrepancy** between the in-sample and out-of-sample R² values points to **overfitting**. The model is too closely fitted to the training data, capturing noise or specific patterns that don’t generalize to unseen data.

__
### Conclusion:

We fitted a linear regression model to the dataset and evaluated its performance both on the training data and test data:

1. **Model Fitting and Evaluation**: The model explained about 46.7% of the variance in the training data, suggesting a reasonable fit to the training data.

2. **Prediction and Comparison**: The out-of-sample R-squared on the test data was only 0.0025, indicating that the model failed to generalize well to unseen data.

3. **Identifying Overfitting**: The large difference between in-sample and out-of-sample R-squared values points to overfitting, where the model fits the training data well but fails to predict new data accurately.

4. **Implications of Overfitting**: The model's poor test performance suggests it has too many parameters or is overly complex, making it ineffective for new data.

5. **Next Steps**: To address overfitting, we could simplify the model, increase the dataset size, or use techniques like regularization and cross-validation.

### Final Thoughts:
Although the model fits the training data well, its poor performance on test data shows a need for adjustments to improve its ability to generalize to new data.

__
## ChatGPT Session Summary
https://chatgpt.com/share/6736cd29-bda4-800e-b950-910bed1501c8

In our exchanges, we explored the process of fitting a linear regression model to a dataset and evaluating its performance using both in-sample and out-of-sample R-squared values. The in-sample R-squared value (0.467) initially suggested that the model explained a decent portion of the variance within the training data. However, when tested on the out-of-sample data, the R-squared value dropped drastically to approximately 0.0025, which indicated that the model was not generalizing well to new, unseen data. This significant disparity between the in-sample and out-of-sample R-squared values is a classic sign of overfitting, where the model is overly complex and captures noise in the training data rather than the true underlying relationships. We discussed potential solutions to improve the model's generalization, such as simplifying the model, utilizing cross-validation to better assess performance, and increasing the size of the training dataset. This analysis highlighted the importance of balancing model complexity and the ability to generalize to ensure better performance on unseen data.

___

# 6. 

In [92]:
# "Cond. No." WAS 343.0 WITHOUT to centering and scaling
model3_fit.summary() 

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     HP   R-squared:                       0.148
Model:                            OLS   Adj. R-squared:                  0.143
Method:                 Least Squares   F-statistic:                     34.40
Date:                Fri, 15 Nov 2024   Prob (F-statistic):           1.66e-14
Time:                        04:27:44   Log-Likelihood:                -1832.6
No. Observations:                 400   AIC:                             3671.
Df Residuals:                     397   BIC:                             3683.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     42.5882      3.580     11.897      0.000      35.551      49.626
Attack         0.2472      0.041      6.051      0.000       0.167       0.327
Defense        0.1001      0.045      2.201      0.028       0.011       0.190
==============================================================================
Omnibus:                      284.299   Durbin-Watson:                   2.006
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             5870.841
Skew:                           2.720   Prob(JB):                         0.00
Kurtosis:                      20.963   Cond. No.                         343.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [93]:
from patsy import center, scale

model3_linear_form_center_scale = \
  'HP ~ scale(center(Attack)) + scale(center(Defense))' 
model_spec3_center_scale = smf.ols(formula=model3_linear_form_center_scale,
                                   data=pokeaman_train)
model3_center_scale_fit = model_spec3_center_scale.fit()
model3_center_scale_fit.summary()
# "Cond. No." is NOW 1.66 due to centering and scaling

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     HP   R-squared:                       0.148
Model:                            OLS   Adj. R-squared:                  0.143
Method:                 Least Squares   F-statistic:                     34.40
Date:                Fri, 15 Nov 2024   Prob (F-statistic):           1.66e-14
Time:                        04:27:50   Log-Likelihood:                -1832.6
No. Observations:                 400   AIC:                             3671.
Df Residuals:                     397   BIC:                             3683.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==========================================================================================
                             coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------
Intercept                 69.3025      1.186     58.439      0.000      66.971      71.634
scale(center(Attack))      8.1099      1.340      6.051      0.000       5.475      10.745
scale(center(Defense))     2.9496      1.340      2.201      0.028       0.315       5.585
==============================================================================
Omnibus:                      284.299   Durbin-Watson:                   2.006
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             5870.841
Skew:                           2.720   Prob(JB):                         0.00
Kurtosis:                      20.963   Cond. No.                         1.66
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [98]:
model4_linear_form_CS = 'HP ~ scale(center(Attack)) * scale(center(Defense))'
model4_linear_form_CS += ' * scale(center(Speed)) * Legendary' 
model4_linear_form_CS += ' * scale(center(Q("Sp. Def"))) * scale(center(Q("Sp. Atk")))'
# Legendary is an indicator, so we don't center and scale that

model4_CS_spec = smf.ols(formula=model4_linear_form_CS, data=pokeaman_train)
model4_CS_fit = model4_CS_spec.fit()
model4_CS_fit.summary().tables[-1]  # Cond. No. is 2,250,000,000,000,000

# The condition number is still bad even after centering and scaling

<class 'statsmodels.iolib.table.SimpleTable'>

In [95]:
# Just as the condition number was very bad to start with
model4_fit.summary().tables[-1]  # Cond. No. is 12,000,000,000,000,000

<class 'statsmodels.iolib.table.SimpleTable'>

## Explanation of Analysis

In `model4`, the linear model is specified with a high level of complexity, adding many predictor variables through interaction terms between centered and scaled features like `Attack`, `Defense`, `Speed`, `Sp. Def`, `Sp. Atk`, and the indicator variable `Legendary`. This design leads to an expanded "design matrix" (`model4_spec.exog`), where each interaction term occupies its own column, increasing the number of variables significantly. This increase in dimensionality introduces severe multicollinearity, which is confirmed by examining the correlation matrix of the design matrix (`np.corrcoef(model4_spec.exog)`) and the extremely high condition number for `model4_spec.exog`. Initially, the condition number is around 12 trillion, and even with centering and scaling, it still rises to 2.25 quadrillion—both indicators of severe multicollinearity.

The presence of such high multicollinearity makes the coefficient estimates unstable, which means they are highly sensitive to small changes in the data. As a result, `model4` is essentially overfitting, picking up on random noise in the training data instead of meaningful patterns that could generalize to new data. This overfitting is reflected in the model’s performance: although the in-sample R-squared (how well the model fits the training data) appears reasonable, the out-of-sample R-squared (how well the model performs on new, unseen data) drops dramatically, indicating that the model cannot generalize.

This sharp difference between the in-sample and out-of-sample R-squared values shows how an overly complex model, like `model4`, is prone to severe overfitting due to high multicollinearity. It ends up capturing coincidental relationships in the training data that do not apply outside of it. By comparison, a simpler model, such as `model3`, which avoids excessive interactions and instead focuses on more reliable, simpler relationships, is able to generalize better and shows more stable performance across different data.